nucypher/blockchain/eth/events.py

```diff
@@ -77,3 +80,44 @@ def __getattr__(self, event_name: str):
     def __iter__(self):
         for event_name in self.names:
             yield self[event_name]
+
+
+class ContractEventsThrottler:
+    """
+    Enables Contract events to be retrieved in batches.
+    """
+    # default to 1000 - smallest default heard about so far (alchemy)
+    DEFAULT_MAX_BLOCKS_PER_CALL = int(os.environ.get(NUCYPHER_EVENTS_THROTTLE_MAX_BLOCKS, 1000))
+
+    def __init__(self,
+                 agent: 'EthereumContractAgent',
+                 event_name: str,
+                 from_block: int,
+                 to_block: int = None,  # defaults to latest block
+                 max_blocks_per_call: int = DEFAULT_MAX_BLOCKS_PER_CALL,
+                 **argument_filters):
+        if not agent:
+            raise ValueError(f"Contract agent must be provided")
+        if not event_name:
+            raise ValueError(f"Event name must be provided")
+
+        self.event_filter = agent.events[event_name]
+        self.from_block = from_block
+        self.to_block = to_block if to_block else agent.blockchain.client.block_number
+        # validity check of block range
+        if to_block <= from_block:
+            raise ValueError(f"Invalid block range provided ({from_block} - {to_block})")
+
+        self.max_blocks_per_call = max_blocks_per_call
+        self.argument_filters = argument_filters
+
+    def __iter__(self):
+        current_from_block = self.from_block
+        current_to_block = min(self.from_block + self.max_blocks_per_call, self.to_block)
+        while current_from_block < current_to_block:
+            for event_record in self.event_filter(from_block=current_from_block,
+                                                  to_block=current_to_block,
+                                                  **self.argument_filters):
+                yield event_record
+            current_from_block = current_to_block
+            current_to_block = min(current_from_block + self.max_blocks_per_call, self.to_block)
```

Comment:
This is a very clever piece of code 😎

Reply:
Some code authors think "clever" is actually an insult. ;-)

Definitely a solid line of code however, core of the concept - perhaps add a comment here :+1:

Reply:
👍  - I've definitely used "clever" with others regarding their code and meant it as a positive. Hopefully, no one was insulted by my use of it 😅

Reply:
Oh my, everything can sound insulting in English